### PR TITLE
tests: Add AddressSanitizer (ASan) Travis build

### PR DESCRIPTION
**Pull request by practicalswift**
Add ASan Travis build.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -102,6 +102,16 @@ jobs:
         RUN_FUNCTIONAL_TESTS=false # Disabled for now. TODO identify suppressions or exclude specific tests
         GOAL="install"
         BITCOIN_CONFIG="--enable-zmq --with-incompatible-bdb --with-gui=qt5 CPPFLAGS=-DDEBUG_LOCKORDER --with-sanitizers=thread --disable-hardening --disable-asm CC=clang CXX=clang++"
+# x86_64 Linux (xenial, no depends, only system libs, sanitizers: address (ASAN))
```
Comment (MarcoFalke) [inline .travis.yml]: Could this be combined with the bionic sanitizer build below instead (assuming that the tests pass)

**Comment by sipa** [comment]
Any reason to not enable ASan and UBSan in the same test?

File: test/functional/p2p_invalid_messages.py [hunk 2]
```diff
@@ -66,7 +67,10 @@ def run_test(self):
         msg_at_size = msg_unrecognized("b" * valid_data_limit)
         assert len(msg_at_size.serialize()) == msg_limit
 
-        with node.assert_memory_usage_stable(perc_increase_allowed=0.03):
+        perc_increase_allowed = 0.03
+        if [s for s in os.environ.get("BITCOIN_CONFIG", "").split(" ") if "--with-sanitizers" in s and "address" in s]:
+            perc_increase_allowed = 3.5
+        with node.assert_memory_usage_stable(perc_increase_allowed=perc_increase_allowed):
```
Comment (MarcoFalke) [inline test/functional/p2p_invalid_messages.py]: I think the test fails unrelated to the sanitizer?

File: .travis/test_04_install.sh [hunk 3]
```diff
@@ -7,9 +7,10 @@
 export LC_ALL=C.UTF-8
 
 travis_retry docker pull "$DOCKER_NAME_TAG"
+export ASAN_OPTIONS="leak_check_at_exit=0"
```
Comment (MarcoFalke) [inline .travis/test_04_install.sh]: Could only suppress these for the gui and wallet?

File: test/functional/p2p_invalid_messages.py (hunk 2 above)
Comment (practicalswift) [reply test/functional/p2p_invalid_messages.py]: I'm not so sure about that: could it be that this failure is intermittent without ASAN and consistent with ASAN? :-) It appears to be consistently triggered under ASAN for me.

**Comment by practicalswift** [comment]
@sipa I was afraid the combined runtime would be too long, but it appears to work. The two Travis jobs have now been merged. Please review.

@MarcoFalke Now running also with LSan (LeakSanitizer). Please review :-)

**Comment by MarcoFalke** [comment]
Could squash everything except the changes in test/functional?

**Comment by practicalswift** [comment]
@MarcoFalke Done! Please re-review :-)

File: test/sanitizer_suppressions/lsan [hunk 4]
```diff
@@ -0,0 +1,7 @@
+leak:addCoin
```
Comment (MarcoFalke) [inline test/sanitizer_suppressions/lsan]: Could add a commit with some documentation about those (see my tsan file in this folder for example):

* `addCoin` is some bench test that doesn't primarily care about leaks.
* The lib* ones are likely upstream false-positives
* ...

**Review by maflcko — APPROVED** [review]

**Comment by practicalswift** [comment]
@MarcoFalke Added comments as suggested. Please re-review :-)

File: test/functional/p2p_invalid_messages.py [hunk 5]
```diff
@@ -66,7 +67,10 @@ def run_test(self):
         msg_at_size = msg_unrecognized("b" * valid_data_limit)
         assert len(msg_at_size.serialize()) == msg_limit
 
-        with node.assert_memory_usage_stable(increase_allowed=0.5):
+        increase_allowed = 0.5
+        if [s for s in os.environ.get("BITCOIN_CONFIG", "").split(" ") if "--with-sanitizers" in s and "address" in s]:
+            increase_allowed = 3.5
+        with node.assert_memory_usage_stable(increase_allowed=increase_allowed):
```
Comment (MarcoFalke) [inline test/functional/p2p_invalid_messages.py]: Could drop this change

File: test/sanitizer_suppressions/lsan [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# addCoin(...) is used in the CoinSelection benchmark
+leak:addCoin
+
+# Suppress LSan warnings triggered in dependencies
+leak:libcrypto
+leak:libqminimal
+leak:libQt5Core
+leak:libQt5Gui
+leak:libQt5Widgets
+leak:new_allocator.h
```
Comment (MarcoFalke) [inline test/sanitizer_suppressions/lsan]: Should suppress the specific benchmark where this is happening, not the gcc internal

**Review by maflcko — DISMISSED** [review]

File: test/sanitizer_suppressions/lsan (hunk 6 above)
Comment (practicalswift) [reply test/sanitizer_suppressions/lsan]: This is the output I'm getting:

```
$ ASAN_SYMBOLIZER_PATH=/usr/lib/llvm-6.0/bin/llvm-symbolizer ASAN_OPTIONS="" LSAN_OPTIONS="suppressions=$(pwd)/test/sanitizer_suppressions/lsan" src/bench/bench_bitcoin -evals=1 -scaling=0
# Benchmark, evals, iterations, total, min, max, median
…

=================================================================
==890==ERROR: LeakSanitizer: detected memory leaks

Indirect leak of 40040 byte(s) in 1001 object(s) allocated from:
    #0 0x535d60 in operator new(unsigned long) (/…/bitcoin/src/bench/bench_bitcoin+0x535d60)
    #1 0x5cd895 in __gnu_cxx::new_allocator<CTxOut>::allocate(unsigned long, void const*) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/ext/new_allocator.h:111:27
    #2 0x5cd895 in std::allocator_traits<std::allocator<CTxOut> >::allocate(std::allocator<CTxOut>&, unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/alloc_traits.h:436
    #3 0x5cd895 in std::_Vector_base<CTxOut, std::allocator<CTxOut> >::_M_allocate(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/stl_vector.h:172
    #4 0x5cd895 in std::vector<CTxOut, std::allocator<CTxOut> >::_M_default_append(unsigned long) /usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/vector.tcc:571

-----------------------------------------------------
Suppressions used:
  count      bytes template
   2002     392392 addCoin
-----------------------------------------------------

SUMMARY: AddressSanitizer: 40040 byte(s) leaked in 1001 allocation(s).
```

Note the absence of the benchmark in the symbolized stack trace.

I agree `leak:new_allocator.h` is not optimal but I don't know what to use instead :-)

Comment (practicalswift) [reply test/sanitizer_suppressions/lsan]: I've now adjusted the suppressions file. Now using `leak:__gnu_cxx::new_allocator<CTxOut>::allocate` which is a bit more specific. Please re-review :-)

Comment (MarcoFalke) [reply test/sanitizer_suppressions/lsan]: Would `leak:src/bench` work?

Comment (practicalswift) [reply test/sanitizer_suppressions/lsan]: @MarcoFalke I think so, but wouldn't that suppression be a bit too broad? Please advice :-)

Comment (practicalswift) [reply test/sanitizer_suppressions/lsan]: Fixed. Please re-review! :-)

**Comment by practicalswift** [comment]
AppVeyor failure seems unrelated.

**Pull request merged**